python_modules/dagster/dagster/_core/definitions/time_window_partitions.py

```diff
@@ -1253,6 +1259,13 @@ class HourlyPartitionsDefinition(TimeWindowPartitionsDefinition):
         # creates partitions (2022-03-12-00:15, 2022-03-12-01:15), (2022-03-12-01:15, 2022-03-12-02:15), ...
     """
 
+    # mapping for fields defined on TimeWindowPartitionsDefinition to this subclasses __new__
+    __field_remap__ = {
+        "start_ts": "start_date",
+        "end_ts": "end_date",
+        "cron_schedule": None,
```

Comment:
the `__new__` on `HourlyPartitionsDefinition` here does not take the `cron_schedule` argument, so we set this to `None` here to prevent trying to pass the `cron_schedule` field from the base class in to this constructor

Reply:
I'm assuming the other __new__ fields (i.e. minute_offset) are passed in through the properties on `TimeWindowPartitionsDefinition`?

Reply:
Im not sure i follow your question, you might need to rephrase it 

An instance of `TimeWindowPartitionsDefinition`, only has the record fields `start_ts`, `timezone`, `end_ts`, `fmt`, `end_offset`, and `cron_schedule` on it, so only those pieces of information are used when we copy or serialize. The other `__new__` arguments like `minute_offset` are available to use when constructing it directly, but a programatic instantiation from an existing instance will only use the record fields.

Reply:
To rephrase the question: I'm assuming that if you call `copy(DailyPartitionsDefinition(...))` then you'll also get another instance of a `DailyPartitionsDefinition`. In that case, how do we pass in arguments to `DailyPartitionsDefinition.__new__` that aren't fields on the `TimeWindowPartitionsDefinition` named tuple (i.e. the `minute_offset` argument)?

Reply:
Though from your comment it sounds like calling `copy(DailyPartitionsDefinition(...))` will return a `TimeWindowPartitionsDefinition` and not a `DailyPartitionsDefinition`

Reply:
ah ya ok so heres the code:
https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_record/__init__.py#L292-L300

so it uses the class of the object it receives and you can do `copy(DailyPartitionsDefinition(...), minute_offset=42)`

Reply:
I see, thanks for explaining. So calling this:
```
partitions_def = HourlyPartitionsDefinition("2023-01-01-00:00", minute_offset=5)
copy(partitions_def)
```
results in a `TimeWindowPartitionsDefinition` without the `minute_offset` set, unless if you provide it explicitly as an argument of copy.

Reply:
It would be nice to handle this but seems out of scope for this PR.

Reply:
yea since `TimeWindowPartitionsDefinition` encodes the `minute_offset` information it receives in `__new__` in to its `cron_schedule` field and `HourlyPartitionsDefinition` does not take `cron_schedule` then that information gets lost on reconstruction. It does return the correct class though. 

```
>>> from dagster import HourlyPartitionsDefinition
>>> from dagster._record import copy
>>> partitions_def = HourlyPartitionsDefinition("2023-01-01-00:00", minute_offset=5)
>>> partitions_def
TimeWindowPartitionsDefinition(start=1672531200.0, end=None, timezone='UTC', fmt='%Y-%m-%d-%H:%M', end_offset=0, cron_schedule='5 * * * *')
>>> partitions_def.__class__
<class 'dagster._core.definitions.time_window_partitions.HourlyPartitionsDefinition'>
>>> c = copy(partitions_def)
>>> c
TimeWindowPartitionsDefinition(start=1672531200.0, end=None, timezone='UTC', fmt='%Y-%m-%d-%H:%M', end_offset=0, cron_schedule='0 * * * *')
>>> c.__class__
<class 'dagster._core.definitions.time_window_partitions.HourlyPartitionsDefinition'>
>>>
```

Reply:
* should these subclasses accept `cron_schedule` hidden via `kwargs` so they round trip correctly? 
* should these subclasses just be functions that return `TimeWindowPartitionsDefinition` ?

Reply:
@OwenKephart  looks like you've had to deal with this problem a bit https://github.com/dagster-io/dagster/blame/master/python_modules/dagster/dagster/_core/definitions/time_window_partitions.py#L2455-L2459 curious if you have thoughts

Reply:
> should these subclasses accept cron_schedule hidden via kwargs so they round trip correctly?

I think yes, we can always unpack the cron schedule to discern the minute/hour/day offset values anyway.

> should these subclasses just be functions that return TimeWindowPartitionsDefinition ?

Not strongly opinionated on this, but that sounds reasonable to me. It would simplify code paths like the one you linked.

Reply:
exploring the latter here: https://github.com/dagster-io/dagster/pull/22988

Reply:
The function idea was fairly dismissed as too crazy.

Got the kwargs approach implemented and put under test. The constructors get a bit more complexity to deal with but not too crazy.